// cdk-time-sleep/lib/time-sleep.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimeSleep = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const fs_1 = require("fs");
const path_1 = require("path");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const aws_lambda_1 = require("aws-cdk-lib/aws-lambda");
const constructs_1 = require("constructs");
class TimeSleep extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const handler = new aws_lambda_1.SingletonFunction(this, 'CustomResourceHandler', {
            // Use raw string to avoid from tightening CDK version requirement
            runtime: new aws_lambda_1.Runtime('nodejs18.x', aws_lambda_1.RuntimeFamily.NODEJS, { supportsInlineCode: true }),
            code: aws_lambda_1.Code.fromInline(fs_1.readFileSync(path_1.join(__dirname, '../', 'lambda', 'dist', 'index.js')).toString()),
            handler: 'index.handler',
            uuid: '494c1b46-0d2f-4e3b-9bfd-0b2cf10162f9',
            lambdaPurpose: 'TimeSleepCustomResourceHandler',
            timeout: aws_cdk_lib_1.Duration.minutes(15),
            memorySize: 128,
        });
        const properties = {
            createDurationSeconds: props.createDuration?.toSeconds() ?? 0,
            destroyDurationSeconds: props.destroyDuration?.toSeconds() ?? 0,
        };
        new aws_cdk_lib_1.CustomResource(this, 'Default', {
            serviceToken: handler.functionArn,
            resourceType: 'Custom::TimeSleep',
            properties,
        });
    }
}
exports.TimeSleep = TimeSleep;
_a = JSII_RTTI_SYMBOL_1;
TimeSleep[_a] = { fqn: "cdk-time-sleep.TimeSleep", version: "0.0.5" };
